All councils page.

whoojemaflip opened the pull request:
A sortable, filterable index page to browse all local authorities. 

https://trello.com/c/6xQryalX/537-update-view-all-councils-page-5

<img width="1174" alt="llm-all-councils-implementation" src="https://cloud.githubusercontent.com/assets/608867/20133602/19ed47ea-a660-11e6-9f85-914963c7ffbf.png">

Missing from the design is the fancy-pants search/filtering widget. There was some feedback from the user tests that I think we wanted to discuss before investing too much time in a build.  This solution uses the Table Filter from `govuk_admin_template` which does the basics for us.  A more sophisticated thing could be built on top of this, once we're sure what to go for.

whoojemaflip committed:
Adds a broken link count field to the Local Authority

To keep complexity out of the query layer and to ensure that these list views
are fast to load, we keep a tally of the number of broken links against each
Local Authority.  This field is recalculated each time a link's status changes.

whoojemaflip committed:
Adds a count of local authorities

whoojemaflip committed:
Adds a 'View All Councils' tab

There are more tabs in the design, but the pages they link to haven't been built yet.

The Home Nav partial can and will be extended with these other links when they appear.

whoojemaflip committed:
Adds a broken link count to each council row

davidbasalla commented on db/migrate/20161104150651_add_broken_link_count_to_local_authority.rb:
```diff
@@ -0,0 +1,11 @@
+  def up
+    add_column :local_authorities, :broken_link_count, :integer, default: 0
+
+    LocalAuthority.all.map(&:calculate_count_of_broken_links)
```
Quick question, is this doing the same as `LocalAuthority.all.each { |la| la.calculate_count_of_broken_links }`? Does using the `map` method have any advantages here?

whoojemaflip replied:
It's exactly the same functionally, but I feel the shorter syntax is a little more intention revealing. It can be a bit confusing the first time you see it though, I agree. Would you rather it were swapped out for the long-hand?

davidbasalla replied:
No it's good like this, was just not used to seeing it.

davidbasalla commented on app/models/local_authority.rb:
```diff
@@ -14,6 +14,13 @@ def provided_services
+  def calculate_count_of_broken_links
+    # to ensure that we can replay our migrations, wrap the update in a guard clause.
+    if respond_to?(:broken_link_count)
+      update_attribute(:broken_link_count, self.links.where.not(status: 200).count)
```
We were wondering if broken links == links.where.not(status: 200) and if there were any other status codes that might count as 'broken', such as redirects. But looking at the link checker it looks like redirects do return a 200 (if final redirects return a 200) so this looks fine. Briefly thought about defining a `broken` scope on the Link model to move the definition of what broken links are to the Link Model... not sure if necessary at this point

whoojemaflip replied:
I really like that. It makes complete sense that when we talk about Broken Links that they be either a subset of all links aka. `Link.all.broken` as you suggest or even a first class model object eg. `class BrokenLink < Link`

davidbasalla commented on spec/models/link_spec.rb:
```diff
@@ -155,4 +155,13 @@
   end
+
+  describe "after_update" do
+    it "triggers a recount of the local authority's broken links" do
```
Would it be worth adding another test here to demonstrate that `calculate_count_of_broken_links` will not get called if `status` does not change?

whoojemaflip replied:
👍

davidbasalla commented on spec/models/local_authority_spec.rb:
```diff
@@ -76,5 +76,13 @@
       end
     end
+
+    describe "calculate_count_of_broken_links" do
```
Small style niggle, I'm used to seeing method tests have a `#` in front of them, see http://betterspecs.org/#describe

davidbasalla replied:
Also the word 'links' is missing from the test description

davidbasalla replied:
We could add another link here with status 200 to show that the broken link count is correct in counting only links that are broken.

whoojemaflip replied:
well spotted. #willfix

davidbasalla replied:
Were also thinking that we wouldn't need to define a local authority as you could use the `local_authority` association from the link factory, but maybe it's more readable? _shrug_

davidbasalla commented on app/views/local_authorities/index.html.erb:
```diff
@@ -10,7 +10,9 @@
       <tr class="table-header">
-        <th>Local Authorities (<%= @authorities.count %>)</th>
+        <th>
+          <div class='count'><span><%= @authorities.count %></span> local authorities</div>
```
Why did you add a "span" element here? Is that good practice generally?

whoojemaflip replied:
I think github ate your markup.. which element are you referring to?

davidbasalla replied:
^ thanks, fixed :) 

So there's a spec which tests the wording that this commit changes. It gets updated in [another commit in this PR](https://github.com/alphagov/local-links-manager/pull/71/commits/c40c642b7615bee8347b3e93521bc662dd37536b). Might be worth either pulling that test change into this commit, or rolling this commit into the other one...

whoojemaflip replied:
the `span` element here gives us the ability to target the numeric bit of the sentence for fancy CSS formatting.

Commit:
Bumps the version of govuk_admin_template

davidbasalla commented on spec/support/appear_before_matcher.rb:
```diff
@@ -0,0 +1,32 @@
+module SortOrderMatchers
```
This is cool! Did you get this pattern from somewhere or is it completely hand rolled? I vaguely remember testing sort order can be a pain...

whoojemaflip replied:
I went looking for it on the internet, and rolled my own based on someone else's code which did 200% of what we needed here. Custom Matchers are the best.

davidbasalla commented on spec/factories/links.rb:
```diff
@@ -5,6 +5,10 @@
     status nil
     link_last_checked nil
+
+    trait :with_service_interaction do
```
This trait is probably not needed because we've already defined the service_interaction association for this factory. We tried removing this trait and any calls to it and the tests still passed.

davidbasalla commented on spec/models/local_authority_spec.rb:
```diff
@@ -76,5 +76,13 @@
+      it "calculates the correct number of broken for the local authority" do
+        local_authority = FactoryGirl.create(:local_authority)
+        FactoryGirl.create(:link, :with_service_interaction, local_authority: local_authority, status: 500)
+        expect { local_authority.calculate_count_of_broken_links }.to change { local_authority.broken_link_count }.from(0).to(1)
```
We can't actually get this test to pass. It appears that when the link instance is being created by FactoryGirl, the after_save callback is already being called so when we get to L84 the broken_link_count is already 1, not 0 as currently expected...

davidbasalla commented on app/views/local_authorities/index.html.erb:
```diff
@@ -19,14 +19,13 @@
-            <%= link_to authority.name, local_authority_services_path(authority.slug), class: 'js-open-on-submit' %>
+          <td>
+            <%= link_to authority.name, local_authority_services_path(authority.slug), class: 'js-open-on-submit local-authority-name' %>
+            <div class="count"><span><%= authority.broken_link_count %> </span>Broken links</div>
```
Could we add a test for this to the integration specs? Not sure how fiddly testing the rows would be...

whoojemaflip committed:
Simplify the LinkStatusUpdater by removing unnecessary indirection.

The `@table` and `@column` indirection were hiding the intention of this
class. By replacing them with the class literals, we make it more
obvious what this class's collaborators are.

whoojemaflip committed:
Rename some class variables and methods for clarity in LinkStatusUpdater

Small refactor to give more descriptive names to things.

davidbasalla commented:
This looks good, we've finished reviewing the commits and left notes accordingly. I didn't look at the CSS too much as it's really my 'fortay' but I'm sure it's good! :)

whoojemaflip committed:
Update LinkStatusUpdater to also update LocalAuthority broken link count

When the status of a Link changes, we should update the Local Authority.
There's no need to do anything particularly clever - it just iterates
over the full list of LAs after the link-checker completes - since the
time-complexity of this routine is entirely bound to the network and
response times of the remote services.

whoojemaflip committed:
Remove the after_save callback on Link#status

ActiveRecord callbacks are a bad idea for a number of reasons. I came
across one of these reasons while trying to test the interaction between
`Links#update` and `LocalAuthority#update_count_of_broken_links` - in
order to get the test working I had to work-around the lifecycle
callback, which hid the true intention of the test and added a lot of
clutter.

whoojemaflip committed:
Adds the ability to Sort the Local Authority list view.

Options are a) number of broken links, and b) alphabetic by name

By default, this view is sorted by the number of broken links.

whoojemaflip committed:
Adds the ability to filter the table by Council Name

This is a simplified version of the filter that was prototyped, as we
felt that the more complex version demanded it's own pull request.

whoojemaflip commented:
@davidbasalla @emmabeynon That was a thorough review! Thanks for pushing me to do better :)

I've rebased the suggested changes into the relevant commits to preserve the history.

As discussed, there's some light refactoring done to the `LinkStatusUpdater` class which allows us to simplify the `Link` and `LocalAuthority` classes from the initial design. Due to a race-hazard, it was a bunch more efficient to drop that refactoring into this PR rather than create several more blocker dependencies.

emmabeynon commented on lib/local-links-manager/check_links/link_status_updater.rb:
```diff
@@ -3,29 +3,27 @@
 module LocalLinksManager
   module CheckLinks
     class LinkStatusUpdater
```
This looks good.  We should probably change similar code in the HomepageStatusUpdater class too so that it's consistent.

emmabeynon replied:
Small change should be made to the commit message so it starts with an active verb.

whoojemaflip replied:
I've just looked at the HomepageStatusUpdater class for the first time. It looks like the updater was maybe planned to be abstracted into common library code at some point, but that hadn't happened yet. I'm not sure what the benefits of consistency are, but I agree that it will need a look at some point.

What with the plans to build a more encompassing Link Checker, do you agree that we should leave this code as-is, for now at least?

brenetic replied:
The one con of that is that the link checker is months away from discovery let alone actually building it. It will come after Lolotron V2 and mid (maybe after, depending . . . ) migration

emmabeynon commented on app/models/link.rb:
```diff
@@ -19,6 +19,11 @@ class Link < ApplicationRecord
 
+  HTTP_OK_STATUS_CODE = 200
+
+  scope :currently_broken, -> { where.not(status: HTTP_OK_STATUS_CODE) }
```
Would it be worth having tests for these new scopes?

whoojemaflip replied:
The code is run by the integration tests, so it seemed like overkill to write slow database-bound unit tests for something like this.

Additionally, unit testing a simple scope like this seems a bit like writing tests for ActiveRecord.  What do you think? I'm comfortable with the current level of coverage for this.

davidbasalla replied:
Sounds good, agreed that the scope is fairly simple so not necessary to test further

emmabeynon commented on app/models/link.rb:
```diff
@@ -1,6 +1,5 @@
 class Link < ApplicationRecord
   before_update :set_time_and_status_on_updated_link, if: :url_changed?
-  after_save :trigger_recount_of_broken_links, if: :status_changed?
```
Something we just thought of is that when an individual link is updated, without this `after_save` callback the broken link recount will not occur, and therefore the broken link count will be incorrect.  We should add some tests for this and figure out how to best trigger this recount for individually edited links.

whoojemaflip replied:
good catch

davidbasalla commented on lib/local-links-manager/check_links/link_status_updater.rb:
```diff
@@ -8,15 +8,29 @@ def initialize(link_checker = LinkChecker.new)
 
       def update
+        check_all_links
+        update_local_authorities_with_broken_link_count
```
Just thinking about this... currently the broken link counts for LAs only get updated after all links have been checked. Currently this is run at 2am every night and takes about 6.5 hours to finish (estimate based on highly unscientific guesswork), so the broken link counts would only be updated around 8:30 am. Are we happy with this sort of delay in the results? Anyone using the service before 8:30 am would see wrong results...

whoojemaflip replied:
maybe we should recalculate after every link check completes?

brenetic commented:
👍
